api/internal/statemachine/event_handler.go

```diff
@@ -0,0 +1,41 @@
+package statemachine
+
+import (
+	"context"
+	"fmt"
+	"runtime/debug"
+	"time"
+)
+
+// eventHandler is a struct that holds the event handler function and its timeout.
+type eventHandler struct {
+	handler EventHandlerFunc
+	timeout time.Duration // Timeout for the handler to complete, default is 5 seconds
+}
+
+// TriggerHandler triggers the event handler for a state transition. The trigger is blocking and will wait for the handler to complete or timeout.
+func (eh *eventHandler) TriggerHandler(ctx context.Context, fromState, toState State) error {
+	ctx, cancel := context.WithTimeout(ctx, eh.timeout)
+	defer cancel()
+	done := make(chan error, 1)
+
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				// Capture panic but don't affect the transition
+				err := fmt.Errorf("event handler panic from %s to %s: %v: %s\n", fromState, toState, r, debug.Stack())
+				done <- err
+			}
+		}()
+		eh.handler(ctx, fromState, toState)
+		close(done)
```

Comment:
i think this should happen in the defer above so it also gets closed on panic

Reply:
Done via - https://github.com/replicatedhq/embedded-cluster/pull/2363/commits/f7e59592be57196c85b6ba9392f574308f669a3b